core/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -222,4 +230,89 @@ public ByteSizeValue paramAsSize(String key, ByteSizeValue defaultValue) {
         return params;
     }
 
+    /**
+     * Does this request have content or a {@code source} parameter? Use this instead of {@link #hasContent()} if this
+     * {@linkplain RestHandler} treats the {@code source} parameter like the body content.
+     */
+    public final boolean hasContentOrSourceParam() {
+        return hasContent() || hasParam("source");
+    }
+
+    /**
+     * The {@linkplain XContentType} of either the request body or the {@code source} param if either exists and can be guessed, otherwise
+     * {@code null}. Only use this if you need to do something different if the request is plain text. Use
+     * {@link #contentOrSourceParamParser()} if you need a parser.
+     */
+    @Nullable
+    public final XContentType contentOrSourceParamXContentType() {
+        BytesReference content = contentOrSource();
+        return content == null ? null : XContentFactory.xContentType(content);
+    }
+
+    /**
+     * A parser for the contents of this request if it has contents, otherwise a parser for the {@code source} parameter if there is one,
+     * otherwise throws an {@link ElasticsearchParseException}. Use {@link #contentOrSourceParamParserOrNull()} instead if you need to
+     * handle the absence of a body gracefully.
+     */
+    public final XContentParser contentOrSourceParamParser() throws IOException {
+        BytesReference content = contentOrSource();
+        if (content.length() == 0) {
+            throw new ElasticsearchParseException("Body required");
+        }
+        return XContentFactory.xContent(content).createParser(content);
+    }
+
+    /**
+     * A parser for the contents of this request if it has contents, otherwise a parser for the {@code source} parameter if there is one,
+     * otherwise {@code null}. Use {@link #contentOrSourceParamParser()} if you should throw an exception back to the user when there isn't
+     * a body. See also {@link #withContentOrSourceParamParserOrNull(IOConsumer)} which may cleaner to use.
+     */
+    @Nullable
+    public final XContentParser contentOrSourceParamParserOrNull() throws IOException {
+        BytesReference content = contentOrSource();
+        if (content.length() == 0) {
+            return null;
+        }
+        return XContentFactory.xContent(content).createParser(content);
+    }
+
+    /**
+     * Helper around {@link #contentOrSourceParamParserOrNull()} that closes the parser for you. This is just syntactic sugar to ease
+     * handling the nullable parser.
+     */
+    @Nullable
+    public final void withContentOrSourceParamParserOrNull(IOConsumer<XContentParser> withParser) throws IOException {
```

Comment:
I'm not really sure how much I like this. It saves a few lines every time but only a few. If we could entirely replace `contentOrSourceParamParserOrNull` with this then I'd like it, otherwise I'm tempted to remove this.